src/cpp/src/storage/checkpointer.cpp

```diff
@@ -16,58 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
-    
-    std::string new_embeddings_file = tmp_checkpoint_dir
-                                        + PathConstants::embeddings_file 
-                                        + PathConstants::file_ext;
-    std::string new_embeddings_state_file = tmp_checkpoint_dir
-                                            + PathConstants::embeddings_state_file
-                                            + PathConstants::file_ext;
-
-    if(std::filesystem::exists(prev_checkpoint_dir)) {
-        std::string embeddings_file = prev_checkpoint_dir 
-                                        + PathConstants::embeddings_file 
-                                        + PathConstants::file_ext;
-        std::string embeddings_state_file = prev_checkpoint_dir 
-                                                + PathConstants::embeddings_state_file
-                                                + PathConstants::file_ext;
-
-        if(std::filesystem::exists(embeddings_file)) {
-            std::filesystem::rename(embeddings_file, new_embeddings_file);
-            std::filesystem::rename(embeddings_state_file, new_embeddings_state_file);
-        }
-    } else {
-        std::string embeddings_file = checkpoint_dir 
-                                        + PathConstants::embeddings_file
-                                        + PathConstants::file_ext;
-        std::string embeddings_state_file = checkpoint_dir 
-                                                + PathConstants::embeddings_state_file
-                                                + PathConstants::file_ext;
-
-        if(std::filesystem::exists(embeddings_file)) {
-            std::filesystem::copy(embeddings_file, new_embeddings_file);
-            std::filesystem::copy(embeddings_state_file, new_embeddings_state_file);
-        }
+
+    std::string new_embeddings_file = tmp_checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    std::string embeddings_file = checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string embeddings_state_file = checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    if (std::filesystem::exists(embeddings_file)) {
+        if (this->config_->save_encoded) std::filesystem::copy(embeddings_file, new_embeddings_file);
+        if (this->config_->save_state) std::filesystem::copy(embeddings_state_file, new_embeddings_state_file);
     }
 
     this->save(tmp_checkpoint_dir, checkpoint_meta);
 
     string final_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "/";
     std::filesystem::rename(tmp_checkpoint_dir, final_checkpoint_dir);
-    
-    std::filesystem::remove_all(prev_checkpoint_dir);
+
+    int delete_checkpoint = epochs - this->config_->save_prev_num * this->config_->interval;
```

Comment:
I think for simplicity his save_prev_num option can be deleted and we just keep around all checkpoints for now. 

In a future PR it would be nice to have a feature called "save_best" which will save the checkpoint which performs best on the valid/test set.